src/pages/experiments/[experimentId]/data-processing/configure-embedding/components/EmbeddingPreview.jsx

```diff
@@ -17,487 +16,261 @@ import PlotStyling from '../../filter-cells/components/PlotStyling';
-          gridOpacity: { value: (config.transGrid / 20) },
-        },
-      ],
-      marks: [
-        {
-          type: 'symbol',
-          from: { data: 'embeddingCat' },
-          encode: {
-            enter: {
-              x: { scale: 'x', field: 'umap1' },
-              y: { scale: 'y', field: 'umap2' },
-              size: { value: config.pointSize },
-              stroke: {
-                scale: 'color',
-                field: 'geneExpression',
-              },
-              fill: {
-                scale: 'color',
-                field: 'geneExpression',
-              },
-              shape: { value: config.pointStyle },
-              fillOpacity: { value: config.pointOpa / 10 },
-            },
-          },
-        },
-
-      ],
-      legends: config.legend,
-      title:
-      {
-        text: { value: config.titleText },
-        color: { value: config.masterColour },
-        anchor: { value: config.titleAnchor },
-        font: { value: config.masterFont },
-        dx: { value: config.bounceX },
-        fontSize: { value: config.titleSize },
-      },
-    };
-  }
-
-  generateData() {
-    const { data } = this.state;
-
-    return data;
-  }
-
-  updatePlotWithChanges(obj) {
-    this.setState((prevState) => {
-      const newState = _.cloneDeep(prevState);
-
-      _.merge(newState.config, obj);
-
-      return newState;
+    const spec = plots[selectedSpec].specGenerator(config);
+    generateData(spec);
+    console.log(config, spec);
+    setPlotSpec(spec);
+  }, [config]);
+
+  useEffect(() => {
+    setConfig(plots[selectedSpec].initialConfig);
+  }, [selectedSpec]);
+
+  // Quick and dirty function to massage prepared data into a good shape.
+  // This will be changed once we actually load data from Redux.
+  const generateData = (spec) => {
+    spec.data.map((s) => {
+      if (s.name === 'cellSets') {
+        s.values = [];
+
+        UMAP.forEach((cell, i) => {
+          s.values[cell.cluster_id] = {
+            name: `${cell.cluster_id}`,
+            cellSetId: cell.cluster_id,
+            cellIds: s.values[cell.cluster_id]?.cellIds ? [...s.values[cell.cluster_id].cellIds, i] : [i],
+            color: colorProvider.getColor(),
+          };
+          // cluster_id
+        });
+      }
+
+      if (s.name === 'expression') {
+        s.values = { expression: UMAP.map((cell) => cell.doubletScore || 0) };
+      }
+
+      if (s.name === 'embedding') {
+        s.values = UMAP.map((cell) => [cell.UMAP_1, cell.UMAP_2]);
+      }
     });
-  }
-
-  changePlot(val) {
-    this.updatePlotWithChanges({ plotToDraw: val });
-    if (val) {
-      this.updatePlotWithChanges({ plotTitle: 'default clusters' });
-    } else {
-      this.updatePlotWithChanges({ plotTitle: 'cell doublet score' });
+  };
+
+  const updatePlotWithChanges = (obj) => {
+    const newConfig = _.cloneDeep(config);
+    _.merge(newConfig, obj);
+    setConfig(newConfig);
+  };
+
+  const renderPlot = () => {
```

Comment:
renderPlot is actually just a helper function to step through the ifs to render the plot. Would it be better to just use a set of ternary operators e.g. `err ? ( <PlatformError /> ) : !config : ( <Spin /> ) : <Vega >` in place of `renderPlot`?

Reply:
I would keep it as is. Chaining ternary operators is not exactly easy to read.